File: algorithms/CPlusPlus/Recursion/linear-search.cpp
/*
Description: linear search using recursion

Time Complexity: O(n) where n is the number of elements in the array
*/

#include <iostream>
#include <vector>
using namespace std;

//function starts
int linearSearch(vector<int> arr, int i, int n, int target)
{
    //base case
    if (i == n)
    {
        return -1;
    }
    //if the ith element in the array is equal to the target, return the index
    if (arr[i] == target)
    {
        return i;
    }
    //recursive function call
    //i refers to the index that we will need to check the element in the array
    //at every recursive call we will increment the i by 1
    return linearSearch(arr, i + 1, n, target);
}

//main starts
int main()
{
    vector<int> arr = {1, 2, 3, 4, 5};
    int n = arr.size();
    int target = 3;
    cout << linearSearch(arr, 0, n, target);
    return 0;
}

/*
Input:
arr = [1,2,3,4,5]
target = 3

Output:
2

*/

File: algorithms/Java/recursion/linear-search.java
package com.dsa;
/*
Description: Linear search in java using recursion

Time Complexity: O (N)
 */

public class LinearSearch {
    //main starts
    public static void main(String[] args) {
        int arr[]={7,2,1,98,43,12,55};
        int target = 2;
        int ans = linearSearch(arr, target, 0);
        System.out.println("Element found at index: "+ ans);
    }

    //function
    static int linearSearch(int[] arr, int target, int i) {
        //base case
        if(i==arr.length){
            return -1;
        }
        //if the target element is found, return its index
        if(arr[i]==target){
            return i;
        }
        //make a recursive function call by incrementing variable i by 1
        return linearSearch(arr, target, i+1);
    }
}
